File: Divide and Conquer/Find Inversions in Array.py
"""Number of inversions.

The code below calculates the number of inversions inside an array in
O(n log(n)) time.
It uses a Divide and Conquer Approach (Merge Sort).
"""


# Combining two subproblems and calculating the inversions
def merge(A, p, q, t):

    l = A[p:q + 1]
    r = A[q + 1:t + 1]

    i, j, inversions = [0, 0, 0]

    for k in range(p, t + 1):
        if i == len(l):
            A[k] = r[j]
            j += 1
        elif j == len(r):
            A[k] = l[i]
            i += 1
        elif l[i] <= r[j]:
            A[k] = l[i]
            i += 1
        else:
            A[k] = r[j]
            j += 1
            inversions += len(l) - i

    return inversions


# Dividing the problem into smaller subproblems
def merge_sort(A, p, t):
    if p < t:
        inversions = 0
        q = int((p + t) / 2)
        inversions += merge_sort(A, p, q)
        inversions += merge_sort(A, q + 1, t)
        inversions += merge(A, p, q, t)
        return inversions
    else:
        return 0


if __name__ == '__main__':
    A = []
    with open('IntegerArray.txt') as f:
        for n in f.read().split():
            A.append(int(n))

    inversions = merge_sort(A, 0, len(A) - 1)
    print("Inversions: ", inversions)

File: Divide and Conquer/Find Inversions in Array.cpp
#include <bits/stdc++.h>
using namespace std;

int merge(int A[], int p, int q, int t) {
	int n1 = q - p + 1;
	int n2 = t - q;
	int l[n1], r[n2];

	int i, j, k, inversions=0;

	for(i=0; i<n1; i++)
		l[i] = A[p+i];
	for(j=0; j<n2; j++)
		r[j] = A[q+1+j];

	for(i = 0, j = 0, k = p; k <= t; k++) {
		if (i == n1) {
            A[k] = r[j++];
        } else if (j == n2) {
            A[k] = l[i++];
        } else if (l[i] <= r[j]) {
            A[k] = l[i++];
        } else {
            A[k] = r[j++];
            inversions += n1 - i;
        }
	}

	return inversions;
}

int merge_sort(int A[], int p, int t) {
	int q;
	if(p < t) {
		int inversions = 0;
		q = (p+t) / 2;
		inversions += merge_sort(A, p, q);
		inversions += merge_sort(A, q+1, t);
		inversions += merge(A, p, q, t);
		return inversions;
	} else
		return 0;
}

int main() {

	int count;
	cout << "Enter size of array: ";
	cin >> count;

	int num[count];
	cout << "Enter the array: ";
	for(int i=0; i<count; i++) {
		cin >> num[i];
	}

	int inversions = merge_sort(num, 0, count-1);
	cout << "Inversions: " << inversions << endl;
}